lib/DetWPLib.thy

```diff
@@ -124,4 +124,67 @@ lemma det_wp_assert_opt :
   apply simp
   done
 
+lemma det_wp_def2:
+  "det_wp P f = (\<forall>s. P s \<longrightarrow> (\<not> snd (f s)) \<and> (\<exists>r. r \<in> fst (f s) \<and> (\<forall>r'. r' \<in> fst (f s) \<longrightarrow> r' = r)))"
+  apply (clarsimp simp: det_wp_def)
+  apply (subst singleton_iff[symmetric])
+  apply (intro iffI conjI impI allI)
+    apply fastforce
+   apply fastforce
+  apply clarsimp
+  apply (drule_tac x=s in spec)
+  apply clarsimp
+  apply (rule_tac x=r in exI)
```

Comment:
these two `rule_tac` need a `rename_tac` before to make the names `s` and `r` stable

Reply:
Yeah I wasn't sure about this sort of case: here Isabelle will generate green variables with the same names that I used in the statement of the lemma, right? Do we still need to `rename_tac` this sort of thing?

Reply:
While it would probably always be fine in this case, the name of those green variables depends on the context and can change unexpectedly, while using `rename_tac` makes it (almost) completely stable. We're not particularly consistent with this in the main proofs but it's more important in `lib` and is a requirement for anything that might be released to the AFP.

Reply:
Yes, they are not guaranteed. It would be fairly stable here, but the Isabelle2025 update for instance merged the name space for constants and free variables, changing a lot of names that previously were thought stable. So the policy is to always explicitly name green variables somewhere.